docs/architecture/authentication-and-authorization.md

```diff
@@ -0,0 +1,125 @@
+
+![](./assets/graphql-security.svg)
+
+When GraphQL operation is processed, an authorization directive is triggered, before actual GraphQL resolver. It will check if the client has required scopes to do the requested operation. To avoid defining permissions statically in GraphQL schema, a YAML file will be loaded with needed requests. In a reality, it will be a ConfigMap injected to Director/Connector. 
+
+The following example illustrates how we can implement dynamic comparison between required scopes and request scopes for example mutations, queries and type fields:
+
+```graphql
+
+type Mutation {
+    createApplication(in: ApplicationInput!): Application! @secureWithScopes(path: "mutations.createApplication")
+    updateApplication(id: ID!, in: ApplicationInput!): Application! @secureWithScopes(path: "mutations.updateApplication")
+    deleteApplication(id: ID!): Application @secureWithScopes(path: "mutations.deleteApplication")
+}
+
+
+type Query {
+    runtimes(filter: [LabelFilter!], first: Int = 100, after: PageCursor): RuntimePage! @secureWithScopes(path: "queries.runtimes")
+    runtime(id: ID!): Runtime @secureWithScopes(path: "queries.runtime")
+}
+
+type Application {
+    id: ID! 
+    name: String!
+    description: String
+    labels(key: String): Labels!
+    status: ApplicationStatus!
+    webhooks: [Webhook!]! @secureWithScopes(path: "types.Application.webhooks")
+    healthCheckURL: String
+    apis(group: String, first: Int = 100, after: PageCursor): APIDefinitionPage! @secureWithScopes(path: "types.Application.apis")
+    eventAPIs(group: String, first: Int = 100, after: PageCursor): EventAPIDefinitionPage! @secureWithScopes(path: "types.Application.eventAPIs")
+    documents(first: Int = 100, after: PageCursor): DocumentPage! @secureWithScopes(path: "types.Application.documents")
+}
+```
+
+The `path` parameter would specify path in YAML file (ConfigMap) with required scopes for a given resolver. For example:
+```yaml
+queries:
+    runtimes: "runtime:list"
+    runtime: "runtime:get"
+mutations:
+    createApplication: "application:create"
+    updateApplication: "application:update"
+    deleteApplication: "application:delete"
+types:
+    Application:
+        webhooks: "webhook:list"
+        apis: "api:list"
+        eventAPIs: "eventapi:list"
+        documents: "document:list"
+```
+The actual scopes will be defined later.
+
+## Flows
+
+Each flow will be set up on a separate host via different VirtualService, as currently OathKeeper doesn't support certificates and multiple `Bearer` authenticators.
+
+### OAuth 2.0
+There are two ways of creating a `client_id` and `client_secret` pair in the Hydra, using Hydra's [oauth client](https://github.com/kyma-project/kyma/blob/ab3d8878d013f8cc34c3f549dfa2f50f06502f14/docs/security/03-06-oauth2-server.md#register-an-oauth2-client) or [simple POST request](https://github.com/kyma-incubator/examples/tree/master/ory-hydra/scenarios/client-credentials#setup-an-oauth2-client).
```

Comment:
when this pair will be created? what if the given pair were compromised, can we reuse client_id?

Reply:
I clarified it after our discussions.